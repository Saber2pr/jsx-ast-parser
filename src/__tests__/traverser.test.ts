/*
 * @Author: saber2pr
 * @Date: 2021-10-02 15:31:36
 * @Last Modified by: saber2pr
 * @Last Modified time: 2021-10-03 09:29:07
 */
import { compiler, parser, transformer, traverser } from '../'

describe('Traverser', () => {
  const code = `
<div 
  width={100} 
  contentEditable={true}
  color={"red"}
  arr={[1,2]}
  style={{
    width: 100,
    color: "red",
    background: 'blue',
    test: {
      color: 'red'
    },
    child: <span>233</span>
  }} 
  id="233ccc" 
  class2Name="qwq123"
  onClick={onClick}
  onError={(error,test) => {
    console.log(error);
    console.log(test)
  }}
>
  <List
    list={[
      {
        content: <View color="red">233</View>,
        logo: <Image mode="test" />
      },
      {
        content: <View />
      }
    ]}
  />
  <div />
  <div id="qwq" />
  <span>aaa</span>
  <span>1234</span>
  <span>1234asd</span>
  <span>
    12
    aa
    <span>aaa</span>
    aa234
    234aaa
  </span>
</div>
  `
  const ast = parser.parse(code)
  const jsx = transformer.transform(ast)

  it('map', () => {
    const jsx2 = traverser.traverse(jsx, node => {
      if (transformer.isJsxElement(node)) {
        if (node.props && node.props.id === 'qwq') {
          return transformer.createJsxElement(node.tagName, {
            ...node.props,
            meta: 233,
          })
        }
      }
    })

    expect(jsx2).toMatchSnapshot()
  })

  it('findNode', () => {
    const node = traverser.findNode(jsx, node => {
      return transformer.isJsxElement(node) && node.tagName === 'List'
    })
    expect(node).toMatchSnapshot()
    expect(compiler.compile(node)).toEqual(
      `[<List list={[{content:<View color="red">233</View>,logo:<Image mode="test"/>},{content:<View/>}]}/>]`
    )

    // get props list source code
    const result = node[0]
    const contents: string[] = []
    if (transformer.isJsxElement(result)) {
      const list = result.props.list
      if (Array.isArray(list)) {
        list.forEach((item: any) =>
          contents.push(compiler.compile(item.content))
        )
      }
    }
    expect(contents).toEqual(['<View color="red">233</View>', '<View/>'])
  })
})
