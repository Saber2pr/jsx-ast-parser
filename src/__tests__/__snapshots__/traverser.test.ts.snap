// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Traverser map 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "props": Object {
        "list": Array [
          Object {
            "content": Object {
              "children": Array [
                Object {
                  "nodeValue": "233",
                  "tagName": "text",
                },
              ],
              "props": Object {
                "color": "red",
              },
              "tagName": "View",
            },
            "logo": Object {
              "children": Array [],
              "props": Object {
                "mode": "test",
              },
              "tagName": "Image",
            },
          },
          Object {
            "content": Object {
              "children": Array [],
              "props": Object {},
              "tagName": "View",
            },
          },
        ],
      },
      "tagName": "List",
    },
    Object {
      "children": Array [],
      "props": Object {},
      "tagName": "div",
    },
    Object {
      "children": Array [],
      "props": Object {
        "id": "qwq",
        "meta": 233,
      },
      "tagName": "div",
    },
    Object {
      "children": Array [
        Object {
          "nodeValue": "aaa",
          "tagName": "text",
        },
      ],
      "props": Object {},
      "tagName": "span",
    },
    Object {
      "children": Array [
        Object {
          "nodeValue": "1234",
          "tagName": "text",
        },
      ],
      "props": Object {},
      "tagName": "span",
    },
    Object {
      "children": Array [
        Object {
          "nodeValue": "1234asd",
          "tagName": "text",
        },
      ],
      "props": Object {},
      "tagName": "span",
    },
    Object {
      "children": Array [
        Object {
          "nodeValue": "12aa",
          "tagName": "text",
        },
        Object {
          "children": Array [
            Object {
              "nodeValue": "aaa",
              "tagName": "text",
            },
          ],
          "props": Object {},
          "tagName": "span",
        },
        Object {
          "nodeValue": "aa234234aaa",
          "tagName": "text",
        },
      ],
      "props": Object {},
      "tagName": "span",
    },
  ],
  "props": Object {
    "class2Name": "qwq123",
    "color": "red",
    "contentEditable": true,
    "id": "233ccc",
    "style": Object {
      "background": "blue",
      "child": Object {
        "children": Array [
          Object {
            "nodeValue": "233",
            "tagName": "text",
          },
        ],
        "props": Object {},
        "tagName": "span",
      },
      "color": "red",
      "test": Object {
        "color": "red",
      },
      "width": 100,
    },
    "width": 100,
  },
  "tagName": "div",
}
`;
